### PR TITLE
UDP Discovery

### PR DESCRIPTION
**Pull request by arkpar**
In this PR:
* Discovery over UPD
* Node table backed by json file
* Node key file
* Public IP address detection

### CODE REVIEW

File: ethcore/src/client.rs [hunk 1]
```diff
@@ -419,11 +419,11 @@ impl BlockChainClient for Client {
 	}
 
 	fn state_data(&self, _hash: &H256) -> Option<Bytes> {
-		unimplemented!();
+		None
```
Comment (arkpar) [inline ethcore/src/client.rs]: Prevents panic when someone requests NodeData from us

File: parity/main.rs [hunk 2]
```diff
@@ -63,8 +64,8 @@ Options:
   -d --db-path PATH        Specify the database & configuration directory path [default: $HOME/.parity]
 
   --no-bootstrap           Don't bother trying to connect to any nodes initially.
-  --listen-address URL     Specify the IP/port on which to listen for peers [default: 0.0.0.0:30304].
-  --public-address URL     Specify the IP/port on which peers may connect [default: 0.0.0.0:30304].
+  --listen-address URL     Specify the IP/port on which to listen for peers.
+  --public-address URL     Specify the IP/port on which peers may connect.
```
Comment (arkpar) [inline parity/main.rs]: Defaults are now determined by the networking layer and not by docopts. 
"0.0.0.0" public address does not make sense.

File: sync/src/chain.rs [hunk 3]
```diff
@@ -82,7 +82,7 @@ const RECEIPTS_PACKET: u8 = 0x10;
 
 const NETWORK_ID: U256 = ONE_U256; //TODO: get this from parent
 
-const CONNECTION_TIMEOUT_SEC: f64 = 30f64;
+const CONNECTION_TIMEOUT_SEC: f64 = 10f64;
```
Comment (arkpar) [inline sync/src/chain.rs]: Large timeout and high import speed leads to block queue underrun and sync stalling when waiting for a unresponsive peer for too long.

File: util/src/network/ip_utils.rs [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+		if sa == ptr::null_mut() { return None; }
+
+		let (addr, _) = match unsafe { *sa }.sa_family as i32 {
+			AF_INET => {
+				let sa: *const sockaddr_in = unsafe { mem::transmute(sa) };
+				let sa = & unsafe { *sa };
+				let (addr, port) = (sa.sin_addr.s_addr, sa.sin_port);
+				(
+					IpAddr::V4(Ipv4Addr::new(
+							(addr & 0x000000FF) as u8,
```
Comment (gavofyork) [inline util/src/network/ip_utils.rs]: 1 too many tabs

File: util/src/network/ip_utils.rs [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+							((addr & 0x0000FF00) >>  8) as u8,
+							((addr & 0x00FF0000) >> 16) as u8,
+							((addr & 0xFF000000) >> 24) as u8,
+							)),
+							port
```
Comment (gavofyork) [inline util/src/network/ip_utils.rs]: two too many tabs

File: util/src/network/host.rs [hunk 6]
```diff
@@ -321,57 +288,98 @@ struct ProtocolTimer {
 /// Root IO handler. Manages protocol handlers, IO timers and network connections.
 pub struct Host<Message> where Message: Send + Sync + Clone {
 	pub info: RwLock<HostInfo>,
-	udp_socket: Mutex<UdpSocket>,
 	tcp_listener: Mutex<TcpListener>,
-	connections: Arc<RwLock<Slab<SharedConnectionEntry>>>,
```
Comment (arkpar) [inline util/src/network/host.rs]: `connections` was split into `handshakes` and `sessions` to simplify code in many places

File: util/src/network/node_table.rs [hunk 7]
```diff
@@ -0,0 +1,412 @@
...
+	fn save(&self) {
+		if let Some(ref path) = self.path {
+			let mut path_buf = PathBuf::from(path);
+			if let Err(e) = fs::create_dir_all(path_buf.as_path()) {
+				warn!("Error creating node table directory: {:?}", e);
+				return;
+			};
+			path_buf.push("nodes.json");
+			let mut json = String::new();
+			json.push_str("{\n");
```
Comment (arkpar) [inline util/src/network/node_table.rs]: The structure here is quite simple so it is much faster to generate json text manually.

File: parity/main.rs (hunk 2 above)
Comment (gavofyork) [inline parity/main.rs]: how does the user know which the default listen port is?

**Comment by arkpar** [comment]
Restored default docopt for `--listen-address`

**Comment by NikVolf** [comment]
discovery.rs is rather small itself
yet the amount of changes in the pr is overhelming
maybe we should target to produce more feature-centric prs ;)

**Comment by gavofyork** [comment]
> maybe we should target to produce more feature-centric prs ;)

+1

**Comment by gavofyork** [comment]
additional tests needed for coveralls to give the go-ahead

**Comment by arkpar** [comment]
I've added as much testing as possible.

**Comment by NikVolf** [comment]
"though coverage is nice when green, it's only metrics and not a goal" (c)

**Pull request merged**